### ISSUE TITLE
projects/core/howto/servers.rst should be properly formatted

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6946]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#6946|
|Type|task|
|Created|2014-02-03 17:52:40Z|
|Branch|https://github.com/twisted/twisted/tree/servers-docs-format-6946|

As with [#6943](https://github.com/twisted/twisted/issues/6943), this is updating the rST sources to be correctly formatted.

<details><summary>Searchable metadata</summary>

```
trac-id__6946 6946
type__task task
reporter__hawkowl hawkowl
priority__normal normal
milestone__Lore_to_Sphinx Lore to Sphinx
branch__branches_servers_docs_format_6946 branches/servers-docs-format-6946
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1391449960000000 1391449960000000
changetime__1393513436000000 1393513436000000
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [41543]) Branching to servers-docs-format-6946.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl removed owner|
|-|-|

Putting this up for review.

Mostly formatting changes and a few fixes for api link shortnames.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_3"></a>|@hynek set owner to @hawkowl|
|-|-|

1.15: double twisted.internet.protocol.Protocol
 1. 20: double twisted.internet.protocol.Factory
 1. 59: Please enclose connectionMade in double backticks like connectionLost in the next line.
 1. 216: not sure if it’s in scope of this ticket, but I would really love if you could smuggle in a
  ```
If all you need is a simple factory that builds a protocol without any additional behavior, Twisted 13.1 added with :api:`twisted.internet.protocol.Factory.forProtocol <Factory.forProtocol>` an even simpler approach.
  ```
 1. The chat session at the end doesn’t get properly formatter, cf to http://twistedmatrix.com/documents/13.2.0/core/howto/servers.html how it *should* like (→ bold).
    I would suggest to use the line-highlight feature of Sphinx:
  ```
Here's a sample transcript of a chat session (text entered by the user is highlighted):

.. code-block:: console
   :emphasize-lines: 1,6,8,10,12

    $ telnet 127.0.0.1 8123
    Trying 127.0.0.1...
    Connected to 127.0.0.1.
    Escape character is '^]'.
    What's your name?
    test
    Name taken, please choose another.
    bob
    Welcome, bob!
    hello
    <alice> hi bob
    twisted makes writing servers so easy!
    <alice> I couldn't agree more
    <carrol> yeah, it's great
  ```
  However, that would require us to tweak the CSS because the highlight color (yellow) is hardly discernible on the source code background (green) and I’m not even color blind.

I’m not sure if 4 is appropriate (but opening a ticket for it seems extreme) and I’m definitely not sure how to fix 5.  Doing it as indicated and opening a ticket on tweaking our CSS seems like a solid way forward but maybe you should discuss it on IRC first.

Re-submit for review afterwards please.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl removed owner|
|-|-|

1-3 are done, I'd like to keep 4 out (as I think it should also have an example, which puts it out of scope proper).

5 is fine - this is rST source formatting, and that's a template problem (also, it won't be the same on rtd, which uses the rtd template).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|


Maybe the code could be updated to be closer to PEP8

```

self.factory.numProtocols + 1

instead of

self.factory.numProtocols+1

```

for 4. I think that at least mentioning the new API is an improvement over the current docs.. and is better to have that little note, than to have nothing. 

For 5 maybe we need a ticket as I don't see any extra markup on those extra line... so is not only a CSS problem, but also depends on HTML.
I can help with the template.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj set owner to @hawkowl|
|-|-|

Thanks hawkowl,

Here are some additional comments:

Notes:
 * No broken links.
 * Doc renders ok  with "make html"

Points:
 1. "asynchronous manner: the protocol" -- looks like that should be a semi-colon, or the following should be a bulleted list.
 2. "where set up of the connection" -- should be setup
 3. "RFC 865" -- could be an RFC link (http://sphinx-doc.org/markup/inline.html#role-rfc)
 4. "The connectionMade event" -- connectionMade should be ``connectionMade``
 5. "Because this is a short example, nothing else has yet started up the Twisted reactor." -- that sentence seems unnecessary and possibly confusing.
 6. "build upon similar lower-level abstraction." -- should be abstractions, I think.
 7. "The most popular in internet protocols is being line-based." -- sounds clumsy. Maybe "Many popular Internet protocols are line-based." instead.
 8. "terminated with a CR-LF combinations"  -- "with CR-LF" not sure about combinations.
 9. "For those cases, there is the LineReceiver protocol" -- link to LineReceiver API
 10. "less popular" -- seems unnecessary to say that
 11. "netstring based protocol and a prefixed-message-length protocol." -- link to those APIs

Please merge after answering or addressing the numbered points above, and
adiroban's earlier suggestions.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl removed owner|
|-|-|

1, 2, 3, 4, 5, and 6 are fixed. I rewrote a bit of 7 and 8 to be nicer. Did 9, 10 and 11, plus I included that note for the new API and the highlighting I didn't do before? I thought I did, but it seems I didn't.

Putting up to review since I rewrote things.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Looks good. Thanks for the changes!

A correction from my previous comment. `emphasize-lines` did work, but I was using the old sphinx version provided by Ubuntu. I have updated the virtual env to latest sphinx and now I see the new html markup for emphasized lines.

Do we have a ticket for CSS changes in the theme or I should create one?

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_9"></a>|@hynek set owner to @hawkowl|
|-|-|

I disagree with the removal of the colon, since that makes total sense: like in this example. ;) But keep it, it’s not worth changing again.

But I don’t like the double-many in

```
*Many* protocols build upon similar lower-level abstractions.

*Many* popular internet protocols are line-based, containing text data terminated by line breaks (commonly CR-LF), rather than containing straight raw data.
```

How about “For example, many popular”?

I just saw in line 126: endpoint.listen ist without () but in the next line reactor.run() is with.  Please add () to listen. And to reactor.stop in the next line.  Maybe even better, link to the functions’ API?

Please fix the things above and feel free to merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

(In [41771]) Merge servers-docs-format-6946: projects/core/howto/servers.rst should be properly formatted

Author: hawkowl
Reviewers: adiroiban, rwall, hynek
Fixes: [#6946](https://github.com/twisted/twisted/issues/6946)

Makes the servers documentation a properly formatted rST file, with some extra content improvements.
